File: app/oracle/deck.py
"""Load the deck and model the Tree spread."""
import json
import os
import random
import time

REPO = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))
CARDS_PATH = os.path.join(REPO, "data", "cards.json")

SPREAD_REALMS = ("roots", "trunk", "branches")

# The Tree spread pulls Root / Trunk / Branch. The twelve Shell cards are the axis itself,
# and cards.json has always promised they "can substitute into any position" — but nothing
# ever implemented it, so a quarter of the printed deck (including The World Turtle, the
# camp's own mascot) could never appear in a reading.
#
# One slot per spread may become a Shell card, at most, so it stays an event rather than a
# flavour. At 1-in-10 a busy night sees a handful — often enough that the turtles witness
# it, rare enough that it means something when it lands.
SHELL_CHANCE = float(os.environ.get("ORACLE_SHELL_CHANCE", "0.10"))


def draw_spread(by_realm, rng=random):
    """Pure chance, one card per slot. Returns (picks, axis_slot|None).

    axis_slot names the slot the Turtle's own axis spoke into, if any.
    """
    picks = {slot: rng.choice(by_realm[slot]) for slot in SPREAD_REALMS}
    axis_slot = None
    if by_realm.get("shell") and rng.random() < SHELL_CHANCE:
        axis_slot = rng.choice(SPREAD_REALMS)
        picks[axis_slot] = rng.choice(by_realm["shell"])
    return picks, axis_slot
SLOT_LABEL = {
    "roots": "what to face",
    "trunk": "where you stand",
    "branches": "what to reach for",
    "shell": "the axis speaks",
}


def load_deck(path=CARDS_PATH):
    """Read cards.json, tolerating a concurrent writer (the art batch) mid-write."""
    last_err = None
    for _ in range(8):
        try:
            with open(path, encoding="utf-8") as f:
                data = json.load(f)
            break
        except (json.JSONDecodeError, ValueError) as e:
            last_err = e
            time.sleep(0.25)
    else:
        raise RuntimeError(f"could not read {path}: {last_err}")
    cards = data["cards"]
    by_realm = {}
    for c in cards:
        by_realm.setdefault(c["realm"], []).append(c)
    return data, cards, by_realm


def card_by_id(cards, cid):
    return next((c for c in cards if c["id"] == cid), None)


def card_payload(c, loc=None):
    """The card as the web clients see it (med/thumb image routes + location)."""
    loc = loc or {}
    med = os.path.join(REPO, "cards/web/med", f"{c['id']}.jpg")
    thumb = os.path.join(REPO, "cards/web/thumb", f"{c['id']}.jpg")
    art = os.path.join(REPO, "cards/art", f"{c['id']}.png")
    return {
        "id": c["id"], "name": c["name"], "realm": c["realm"],
        "slot": SLOT_LABEL.get(c["realm"], ""),
        "reading": c["reading"], "shadow": c.get("shadow", ""),
        "turtle_dare": c["turtle_dare"],
        "real_2026": c["real_2026"]["name"],
        "location": {"directions": loc.get("directions", ""), "status": loc.get("status", ""),
                     "clock": loc.get("clock")},
        "image": (f"/med/{c['id']}.jpg" if os.path.exists(med)
                  else (f"/art/{c['id']}.png" if os.path.exists(art) else None)),
        "thumb": (f"/thumb/{c['id']}.jpg" if os.path.exists(thumb)
                  else (f"/med/{c['id']}.jpg" if os.path.exists(med) else None)),
    }

File: cloud/src/deck.js
/* Port of app/oracle/deck.py — the deck and the Tree spread.
 *
 * The Python version reads data/cards.json off disk with a retry loop (the art batch
 * may be mid-write). A Worker has no disk: the deck is bundled at build time, so it
 * is always whole and the retry loop disappears.
 */
import CARDS_DATA from "../../data/cards.json" with { type: "json" };
import { rand, choice } from "./util.js";

export const SPREAD_REALMS = ["roots", "trunk", "branches"];

export const SLOT_LABEL = {
  roots: "what to face",
  trunk: "where you stand",
  branches: "what to reach for",
  shell: "the axis speaks",
};

export const DECK = CARDS_DATA;
export const CARDS = CARDS_DATA.cards;

export const BY_REALM = (() => {
  const out = {};
  for (const c of CARDS) (out[c.realm] = out[c.realm] || []).push(c);
  return out;
})();

/** Pure chance, one card per slot. Returns {picks, axisSlot|null}. */
export function drawSpread(byRealm, shellChance) {
  const picks = {};
  for (const slot of SPREAD_REALMS) picks[slot] = choice(byRealm[slot]);
  let axisSlot = null;
  if ((byRealm.shell || []).length && rand() < shellChance) {
    axisSlot = choice(SPREAD_REALMS);
    picks[axisSlot] = choice(byRealm.shell);
  }
  return { picks, axisSlot };
}

/* Every card in cards/web/{med,thumb} is built by tools/webimg.py for the whole deck,
 * and cloud/prepare-assets.sh copies all of them. So unlike the Python server — which
 * stats each file and degrades to /art/*.png — the cloud paths are unconditional. */
export function cardPayload(c, loc) {
  loc = loc || {};
  return {
    id: c.id,
    name: c.name,
    realm: c.realm,
    slot: SLOT_LABEL[c.realm] || "",
    reading: c.reading,
    shadow: c.shadow || "",
    turtle_dare: c.turtle_dare,
    real_2026: c.real_2026.name,
    location: {
      directions: loc.directions || "",
      status: loc.status || "",
      clock: loc.clock === undefined ? null : loc.clock,
    },
    image: `/med/${c.id}.jpg`,
    thumb: `/thumb/${c.id}.jpg`,
  };
}
